server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java

```diff
@@ -0,0 +1,297 @@
+import org.elasticsearch.search.SearchShardTarget;
+import org.elasticsearch.search.aggregations.InternalAggregation.ReduceContext;
+import org.elasticsearch.search.aggregations.InternalAggregations;
+import org.elasticsearch.search.internal.InternalSearchResponse;
+import org.elasticsearch.search.profile.ProfileShardResult;
+import org.elasticsearch.search.profile.SearchProfileShardResults;
+import org.elasticsearch.search.suggest.Suggest;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Function;
+
+import static org.elasticsearch.action.search.SearchResponse.Clusters;
+
+/**
+ * Merges multiple search responses into one. Used in cross-cluster search when reduction is performed locally on each cluster.
+ * The CCS coordinating node sends one search request per remote cluster involved and gets one search response back from each one of them.
+ * Such responses contain all the info to be able to perform an additional reduction and return results back to the user.
+ * Preconditions are that only non final reduction has been performed on each cluster, meaning that buckets have not been pruned locally
+ * and pipeline aggregations have not yet been executed. Also, from+size search hits need to be requested to each cluster.
+ */
+//TODO it may make sense to investigate reusing existing merge code in SearchPhaseController#reducedQueryPhase, the logic is similar
+//yet there are substantial differences in terms of the objects exchanged and logic in the sortDocs method.
+final class SearchResponseMerger {
+    private final int from;
+    private final int size;
+    private final SearchTimeProvider searchTimeProvider;
+    private final Clusters clusters;
+    private final Function<Boolean, ReduceContext> reduceContextFunction;
+    private final List<SearchResponse> searchResponses = new CopyOnWriteArrayList<>();
+
+    SearchResponseMerger(int from, int size, SearchTimeProvider searchTimeProvider, Clusters clusters,
+                         Function<Boolean, ReduceContext> reduceContextFunction) {
+        this.from = from;
+        this.size = size;
+        this.searchTimeProvider = Objects.requireNonNull(searchTimeProvider);
+        this.clusters = Objects.requireNonNull(clusters);
+        this.reduceContextFunction = Objects.requireNonNull(reduceContextFunction);
+    }
+
+    /**
+     * Add a search response to the list of responses to be merged together into one.
+     * Merges currently happen at once when all responses are available and {@link #getMergedResponse()} is called. That may change
+     * in the future as it's possible to introduce incremental merges as responses come in if necessary.
```

Comment:
I don't think incremental merges is appealing here. The number of remote clusters should be low so there is no benefit to do the reduce incrementally.

Reply:
I think we may come back to this and have a laugh one day about the "number of remote clusters should be low" assumption :) but I agree we are not talking about hundreds at the moment. I thought it may be useful to incrementally reduce given the size of the responses from each cluster, but we should first measure what the benefit is if any.